pkg/stub/handler.go

```diff
@@ -2,96 +2,410 @@ package stub
 
 import (
 	"context"
+	"encoding/json"
+	"fmt"
 	"io/ioutil"
+	"os"
+	"strings"
 
 	"github.com/openshift/cluster-samples-operator/pkg/apis/samplesoperator/v1alpha1"
 
 	"github.com/operator-framework/operator-sdk/pkg/sdk"
 	"github.com/sirupsen/logrus"
+
 	corev1 "k8s.io/api/core/v1"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+
+	kapis "k8s.io/apimachinery/pkg/apis/meta/v1"
 
-	//"k8s.io/apimachinery/pkg/api/errors"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	//"k8s.io/apimachinery/pkg/runtime/schema"
+
 	restclient "k8s.io/client-go/rest"
 
+	imagev1 "github.com/openshift/api/image/v1"
+	templatev1 "github.com/openshift/api/template/v1"
+
 	imagev1client "github.com/openshift/client-go/image/clientset/versioned/typed/image/v1"
 	templatev1client "github.com/openshift/client-go/template/clientset/versioned/typed/template/v1"
 )
 
+const (
+	x86OfficialContentRootDir    = "/opt/openshift/arch/x86_64/official"
+	x86CommunityContentRootDir   = "/opt/openshift/arch/x86_64/community"
+	ppc64OfficialContentRootDir  = "/opt/openshift/arch/ppc64le/official"
+	ppc64CommunityContentRootDir = "/opt/openshift/arch/ppc64le/community"
+	x86                          = "x86_64"
+	ppc                          = "ppc64le"
+)
+
 func NewHandler() sdk.Handler {
-	return &Handler{}
+	h := Handler{}
+
+	h.initter = &defaultInClusterInitter{h: &h}
+	h.initter.init()
+
+	h.sdkwrapper = &defaultSDKWrapper{h: &h}
+
+	h.fileimagegetter = &defaultImageStreamFromFileGetter{h: &h}
+	h.filetemplategetter = &defaultTemplateFromFileGetter{h: &h}
+	h.filefinder = &defaultResourceFileLister{h: &h}
+
+	h.imageclientwrapper = &defaultImageStreamClientWrapper{h: &h}
+	h.templateclientwrapper = &defaultTemplateClientWrapper{h: &h}
+	h.secreteclientwrapper = &defaultSecretClientWrapper{h: &h}
+
+	h.namespace = getNamespace()
+
+	h.skippedImagestreams = make(map[string]bool)
+	h.skippedTemplates = make(map[string]bool)
+
+	return &h
 }
 
 type Handler struct {
-	// Fill me
+	initter InClusterInitter
+
+	sdkwrapper SDKWrapper
+
+	operatorCfg    *v1alpha1.SamplesResource
+	registrySecret *corev1.Secret
+
+	restconfig  *restclient.Config
+	tempclient  *templatev1client.TemplateV1Client
+	imageclient *imagev1client.ImageV1Client
+	coreclient  *corev1client.CoreV1Client
+
+	imageclientwrapper    ImageStreamClientWrapper
+	templateclientwrapper TemplateClientWrapper
+	secreteclientwrapper  SecretClientWrapper
+
+	fileimagegetter    ImageStreamFromFileGetter
+	filetemplategetter TemplateFromFileGetter
+	filefinder         ResourceFileLister
+
+	namespace string
+
+	skippedTemplates    map[string]bool
+	skippedImagestreams map[string]bool
+}
+
+func (h *Handler) manageDockerCfgSecret(deleted bool) error {
+	err := h.secreteclientwrapper.Delete("openshift", h.registrySecret.Name, &metav1.DeleteOptions{})
```

Comment:
I think this needs a // TODO to not just always delete the secret before creating/updating it.

ideally we'd only delete it when the operator's secret is deleted.

Reply:
Since there is not "Status" with the secret, there wasn't any data to preserve in my mind between changes.

So rather than attempting to retrieve (an RPC), then creating or updating based on the results, seemed cumbersome.

I thought a `replace --force` was appropriate.  go-client does not wrapper that semantic in an API call.  So I implemented it manually.

Convinced?

Reply:
my primary concern here is we're creating a window where the secret doesn't exist and imports can fail.

Reply:
(or worse, we successfully delete the old secret and fail to create the new one)